go/pkg/io/impath/impath.go

```diff
@@ -0,0 +1,152 @@
+func (p Absolute) String() string {
+	pstr := string(p.path)
+	if pstr == "" {
+		return Root
+	}
+	return pstr
+}
+
+// String implements the Stringer interface and returns the path as a string.
+func (p Relative) String() string {
+	return string(p.path)
+}
+
+// Append is a convenient method to join additional elements to this path.
+func (p Absolute) Append(elements ...Relative) Absolute {
+	paths := make([]string, len(elements)+1)
+	paths[0] = p.String()
+	for i, p := range elements {
+		paths[i+1] = p.String()
+	}
+	return Absolute{path: filepath.Join(paths...)}
+}
+
+// Append is a convenient method to join additional elements to this path.
+func (p Relative) Append(elements ...Relative) Relative {
+	paths := make([]string, len(elements)+1)
+	paths[0] = p.String()
+	for i, p := range elements {
+		paths[i+1] = p.String()
+	}
+	return Relative{path: filepath.Join(paths...)}
+}
+
+var (
+	zeroAbs = Absolute{}
+	zeroRel = Relative{}
+)
+
+// Abs creates a new absolute and clean path from the specified elements.
+//
+// If the specified elements do not join to a valid absolute path, ErrNotAbsolute is returned.
+func Abs(elements ...string) (Absolute, error) {
+	p := filepath.Join(elements...)
+	if filepath.IsAbs(p) {
+		return Absolute{path: p}, nil
+	}
+	return zeroAbs, errors.Join(ErrNotAbsolute, fmt.Errorf("path %q", p))
+}
+
+// MustAbs is a convenient wrapper of ToAbs that is useful for constant paths.
+//
+// If the specified elements do not join to a valid absolute path, this function panics.
+func MustAbs(elements ...string) Absolute {
+	p, err := Abs(elements...)
+	if err != nil {
+		panic(err)
+	}
+	return p
+}
+
+// Rel creates a new relative and clean path from the specified elements.
+//
+// If the specified elements do not join to a valid relative path, ErrNotRelative is returned.
+func Rel(elements ...string) (Relative, error) {
+	p := filepath.Join(elements...)
+	if filepath.IsAbs(p) {
+		return zeroRel, errors.Join(ErrNotRelative, fmt.Errorf("path %q", p))
+	}
+	return Relative{path: p}, nil
+}
+
+// MustRel is a convenient wrapper of ToRel that is useful for constant paths.
+//
+// If the specified elements do not join to a valid relative path, this function panics.
+func MustRel(elements ...string) Relative {
+	p, err := Rel(elements...)
+	if err != nil {
+		panic(err)
+	}
+	return p
+}
+
+// Descendant returns a relative path to the specified base path such that
+// when joined together with the base using filepath.Join(base, path), the result
+// is lexically equivalent to the specified target path.
+//
+// The returned error is nil, ErrNotRelative, or ErrNotDescendant.
+func Descendant(base Absolute, target Absolute) (Relative, error) {
+	p, err := filepath.Rel(base.String(), target.String())
+	if err != nil {
+		// Since the zero Absolute path is the root, this error may only happen if base and target
+		// have different roots, which is possible on Windows.
+		return zeroRel, errors.Join(ErrNotRelative, err)
```

Comment:
why the error type is `ErrNotRelative` and not `ErrNotDescendant`. If I got the comment right, this can happen if both are absolute paths but belong to different system drives.

Reply:
At this point the function has only tested for a relative path, which is a superset of a descendant path. Returning that error conveys more information to the caller. I.e. a non-descendant path may still be relative, but a non-relative path cannot be descendant.

Reply:
Oh, so you mean here that path `target` cannot be made relative to `base`, is that right? If yes, can you provide the explanation in the function's comment? The error's message mentions `path is not relative` and it's easy to for the users to assume to it's due to arguments not being relative.

Reply:
Will do. The returned error from `filepath.Rel` includes that information. The combined errors would lool like:
```
path is not relative
Rel: can't make baz relative to /foo/bar
```